synapse/storage/data_stores/main/admin.py

```diff
@@ -0,0 +1,142 @@
+    NOT_YET_ACTIVE = NamedConstant()
+    NON_EXISTANT = NamedConstant()
+
+
+@attr.s
+class TokenLookupResult(object):
+    admin_token = attr.ib()
+    permissions = attr.ib(
+        default=attr.Factory(lambda: defaultdict(lambda: defaultdict(lambda: False)))
+    )
+    token_state = attr.ib(type=TokenState, default=TokenState.NON_EXISTANT)
+
+
+class AdminTokenWorkerStore(SQLBaseStore):
+    async def get_permissions_for_token(self, admin_token: str) -> TokenLookupResult:
+        """
+        Get a token and its permissions.
+
+        Args:
+            admin_token: The admin token to look up permissions for.
+        """
+        result = TokenLookupResult(admin_token=admin_token)
+
+        token = await self._simple_select_one(
+            "admin_tokens",
+            {"admin_token": admin_token},
+            ["admin_token", "valid_from", "valid_until"],
+            allow_none=True,
+        )
+
+        # Token does not exist
+        if token is None:
+            result.token_state = TokenState.NON_EXISTANT
+            return result
+
+        elif token["valid_from"] > self.hs.get_reactor().seconds():
+            result.token_state = TokenState.NOT_YET_ACTIVE
+            return result
+
+        # Token is expired
+        elif token["valid_until"] < self.hs.get_reactor().seconds():
+            result.token_state = TokenState.EXPIRED
+            return result
+
+        permissions = await self._simple_select_list(
+            "admin_token_permissions",
+            {"admin_token": token["admin_token"]},
+            ["endpoint", "action", "allowed"],
+        )
+
+        for permission in permissions:
+            if permission["allowed"]:
+                result.permissions[permission["endpoint"]][permission["action"]] = True
+
+        result.token_state = TokenState.VALID
+        return result
+
+    async def create_admin_token(
+        self, valid_until: int, creator: str, description: str
+    ) -> str:
+        """
+        Create a new admin token and return it.
+
+        Args:
+            valid_until: Token's end validity, in seconds since the epoch.
+            creator: mxid of the creating user, if applicable.
+            description: A description of the token's use.
+
+        Returns:
+            The admin token.
+        """
+        now = int(self.hs.get_reactor().seconds())
+
+        if valid_until < now:
+            raise ValueError("Can't end validity in the past")
+
+        token_value = self.hs.get_secrets().token_bytes(16)
+        token_uuid = str(uuid.UUID(bytes=token_value))
+
+        await self._simple_insert(
+            "admin_tokens",
+            {
+                "admin_token": token_uuid,
+                "valid_from": now,
+                "valid_until": valid_until,
+                "created_by": creator,
+                "description": description,
+            },
+        )
+
+        return token_uuid
+
+    async def set_permission_for_token(
```

Comment:
Remember that `DataStore` is a steaming pile of inheritance. To distinguish from other types of token:

```suggestion
    async def set_permission_for_admin_token(
```

(likewise for other methods in this class).